diff -rcN src/newlib_org/config.sub src/newlib/config.sub
*** src/newlib_org/config.sub	2010-07-16 18:26:56.000000000 +0200
--- src/newlib/config.sub	2010-07-16 18:29:09.000000000 +0200
***************
*** 1203,1209 ****
  	      | -*vms* | -sco* | -esix* | -isc* | -aix* | -sunos | -sunos[34]*\
  	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -solaris* | -sym* \
  	      | -amigaos* | -amigados* | -msdos* | -newsos* | -unicos* | -aof* \
! 	      | -aos* \
  	      | -nindy* | -vxsim* | -vxworks* | -ebmon* | -hms* | -mvs* \
  	      | -clix* | -riscos* | -uniplus* | -iris* | -rtu* | -xenix* \
  	      | -hiux* | -386bsd* | -knetbsd* | -mirbsd* | -netbsd* \
--- 1203,1209 ----
  	      | -*vms* | -sco* | -esix* | -isc* | -aix* | -sunos | -sunos[34]*\
  	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -solaris* | -sym* \
  	      | -amigaos* | -amigados* | -msdos* | -newsos* | -unicos* | -aof* \
! 	      | -aos* | -escape* \
  	      | -nindy* | -vxsim* | -vxworks* | -ebmon* | -hms* | -mvs* \
  	      | -clix* | -riscos* | -uniplus* | -iris* | -rtu* | -xenix* \
  	      | -hiux* | -386bsd* | -knetbsd* | -mirbsd* | -netbsd* \
diff -rcN src/newlib_org/newlib/configure.host src/newlib/newlib/configure.host
*** src/newlib_org/newlib/configure.host	2010-07-16 18:26:56.000000000 +0200
--- src/newlib/newlib/configure.host	2010-07-16 18:30:39.000000000 +0200
***************
*** 316,321 ****
--- 316,324 ----
  # THIS TABLE IS ALPHA SORTED.  KEEP IT THAT WAY.
  
  case "${host}" in
+ 	i[3-7]86-*-escape*)
+ 	sys_dir=escape
+ 	;;
    *-*-cygwin*)
  	posix_dir=posix
  	stdio64_dir=stdio64
diff -rcN src/newlib_org/newlib/libc/sys/configure.in src/newlib/newlib/libc/sys/configure.in
*** src/newlib_org/newlib/libc/sys/configure.in	2010-07-16 18:26:56.000000000 +0200
--- src/newlib/newlib/libc/sys/configure.in	2010-07-16 18:32:05.000000000 +0200
***************
*** 20,25 ****
--- 20,26 ----
  
  if test -n "${sys_dir}"; then
    case ${sys_dir} in
+   escape) AC_CONFIG_SUBDIRS(escape) ;;
  	a29khif) AC_CONFIG_SUBDIRS(a29khif) ;;
  	arc) AC_CONFIG_SUBDIRS(arc) ;;
  	arm) AC_CONFIG_SUBDIRS(arm) ;;
