diff -rcN src_org/newlib//config.sub src/newlib//config.sub
*** src_org/newlib//config.sub	2006-12-11 05:12:46.000000000 +0100
--- src/newlib//config.sub	2011-05-31 13:43:18.000000000 +0200
***************
*** 530,535 ****
--- 530,539 ----
  		basic_machine=a29k-amd
  		os=-ebmon
  		;;
+ 	eco32-*)
+ 		basic_machine=eco32
+ 		os=-elf-escape
+ 		;;
  	elxsi)
  		basic_machine=elxsi-elxsi
  		os=-bsd
***************
*** 1203,1209 ****
  	      | -*vms* | -sco* | -esix* | -isc* | -aix* | -sunos | -sunos[34]*\
  	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -solaris* | -sym* \
  	      | -amigaos* | -amigados* | -msdos* | -newsos* | -unicos* | -aof* \
! 	      | -aos* \
  	      | -nindy* | -vxsim* | -vxworks* | -ebmon* | -hms* | -mvs* \
  	      | -clix* | -riscos* | -uniplus* | -iris* | -rtu* | -xenix* \
  	      | -hiux* | -386bsd* | -knetbsd* | -mirbsd* | -netbsd* \
--- 1206,1212 ----
  	      | -*vms* | -sco* | -esix* | -isc* | -aix* | -sunos | -sunos[34]*\
  	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -solaris* | -sym* \
  	      | -amigaos* | -amigados* | -msdos* | -newsos* | -unicos* | -aof* \
! 	      | -aos* | -escape* \
  	      | -nindy* | -vxsim* | -vxworks* | -ebmon* | -hms* | -mvs* \
  	      | -clix* | -riscos* | -uniplus* | -iris* | -rtu* | -xenix* \
  	      | -hiux* | -386bsd* | -knetbsd* | -mirbsd* | -netbsd* \
diff -rcN src_org/newlib/newlib/configure.host src/newlib/newlib/configure.host
*** src_org/newlib/newlib/configure.host	2006-12-18 22:49:14.000000000 +0100
--- src/newlib/newlib/configure.host	2011-05-31 14:02:32.000000000 +0200
***************
*** 114,119 ****
--- 114,121 ----
    d30v*)
  	machine_dir=d30v
  	;;
+   eco32*)
+   	;;
    ep9312)
  	machine_dir=arm
  	;;
***************
*** 353,358 ****
--- 355,363 ----
    d30v*)
  	sys_dir=
  	;;
+   eco32-*-escape*)
+     sys_dir=escape
+     ;;
    ep9312-*-*)
  	sys_dir=arm
  	if [ "x${newlib_may_supply_syscalls}" = "xno" ] ; then
diff -rcN src_org/newlib//newlib/libc/sys/configure.in src/newlib//newlib/libc/sys/configure.in
*** src_org/newlib//newlib/libc/sys/configure.in	1970-01-01 01:00:00.000000000 +0100
--- src/newlib//newlib/libc/sys/configure.in	2011-05-31 13:35:18.000000000 +0200
***************
*** 20,25 ****
--- 20,26 ----
  
  if test -n "${sys_dir}"; then
    case ${sys_dir} in
+   escape) AC_CONFIG_SUBDIRS(escape) ;;
  	a29khif) AC_CONFIG_SUBDIRS(a29khif) ;;
  	arc) AC_CONFIG_SUBDIRS(arc) ;;
  	arm) AC_CONFIG_SUBDIRS(arm) ;;
diff -rcN src_org/newlib/newlib/libc/include/machine/ieeefp.h src/newlib/newlib/libc/include/machine/ieeefp.h
*** src_org/newlib/newlib/libc/include/machine/ieeefp.h	2006-11-08 20:26:43.000000000 +0100
--- src/newlib/newlib/libc/include/machine/ieeefp.h	2011-05-31 14:09:40.000000000 +0200
***************
*** 154,159 ****
--- 154,163 ----
  #define __IEEE_BIG_ENDIAN
  #endif
  
+ #ifdef __eco32__
+ #define __IEEE_BIG_ENDIAN
+ #endif
+ 
  #ifdef __D30V__
  #define __IEEE_BIG_ENDIAN
  #endif
