diff -rcN src/binutils-org/bfd/config.bfd src/binutils/bfd/config.bfd
*** src/binutils-org/bfd/config.bfd	2012-09-04 16:14:59.000000000 +0200
--- src/binutils/bfd/config.bfd	2013-12-01 18:33:52.746392191 +0100
***************
*** 522,527 ****
--- 522,531 ----
    i[3-7]86-*-chorus*)
      targ_defvec=bfd_elf32_i386_vec
      ;;
+   i[3-7]86-*-escape*)
+     targ_defvec=bfd_elf32_i386_vec
+     targ_selvecs=bfd_elf32_i386_vec
+     ;;
    i[3-7]86-*-dicos*)
      targ_defvec=bfd_elf32_i386_vec
      targ64_selvecs="bfd_elf64_x86_64_vec bfd_elf64_l1om_vec bfd_elf64_k1om_vec"
diff -rcN src/binutils-org/bfd/doc/bfd.texinfo src/binutils/bfd/doc/bfd.texinfo
*** src/binutils-org/bfd/doc/bfd.texinfo	2010-10-28 13:40:25.000000000 +0200
--- src/binutils/bfd/doc/bfd.texinfo	2013-12-01 18:37:09.918390620 +0100
***************
*** 322,328 ****
  @printindex cp
  
  @tex
! % I think something like @colophon should be in texinfo.  In the
  % meantime:
  \long\def\colophon{\hbox to0pt{}\vfill
  \centerline{The body of this manual is set in}
--- 322,328 ----
  @printindex cp
  
  @tex
! % I think something like @@colophon should be in texinfo.  In the
  % meantime:
  \long\def\colophon{\hbox to0pt{}\vfill
  \centerline{The body of this manual is set in}
***************
*** 333,339 ****
  \centerline{{\sl\fontname\tensl\/}}
  \centerline{are used for emphasis.}\vfill}
  \page\colophon
! % Blame: doc@cygnus.com, 28mar91.
  @end tex
  
  @bye
--- 333,339 ----
  \centerline{{\sl\fontname\tensl\/}}
  \centerline{are used for emphasis.}\vfill}
  \page\colophon
! % Blame: doc@@cygnus.com, 28mar91.
  @end tex
  
  @bye
diff -rcN src/binutils-org/bfd/elf32-i386.c src/binutils/bfd/elf32-i386.c
*** src/binutils-org/bfd/elf32-i386.c	2013-03-25 09:06:19.000000000 +0100
--- src/binutils/bfd/elf32-i386.c	2013-12-01 18:34:15.530392009 +0100
***************
*** 518,524 ****
  /* The name of the dynamic interpreter.  This is put in the .interp
     section.  */
  
! #define ELF_DYNAMIC_INTERPRETER "/usr/lib/libc.so.1"
  
  /* If ELIMINATE_COPY_RELOCS is non-zero, the linker will try to avoid
     copying dynamic variables from a shared lib into an app's dynbss
--- 518,524 ----
  /* The name of the dynamic interpreter.  This is put in the .interp
     section.  */
  
! #define ELF_DYNAMIC_INTERPRETER "/bin/dynlink"
  
  /* If ELIMINATE_COPY_RELOCS is non-zero, the linker will try to avoid
     copying dynamic variables from a shared lib into an app's dynbss
diff -rcN src/binutils-org/config.sub src/binutils/config.sub
*** src/binutils-org/config.sub	2012-04-25 17:53:25.000000000 +0200
--- src/binutils/config.sub	2013-12-01 18:34:37.014391838 +0100
***************
*** 1348,1354 ****
  	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -auroraux* | -solaris* \
  	      | -sym* | -kopensolaris* \
  	      | -amigaos* | -amigados* | -msdos* | -newsos* | -unicos* | -aof* \
! 	      | -aos* | -aros* \
  	      | -nindy* | -vxsim* | -vxworks* | -ebmon* | -hms* | -mvs* \
  	      | -clix* | -riscos* | -uniplus* | -iris* | -rtu* | -xenix* \
  	      | -hiux* | -386bsd* | -knetbsd* | -mirbsd* | -netbsd* \
--- 1348,1354 ----
  	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -auroraux* | -solaris* \
  	      | -sym* | -kopensolaris* \
  	      | -amigaos* | -amigados* | -msdos* | -newsos* | -unicos* | -aof* \
! 	      | -aos* | -aros* | -escape* \
  	      | -nindy* | -vxsim* | -vxworks* | -ebmon* | -hms* | -mvs* \
  	      | -clix* | -riscos* | -uniplus* | -iris* | -rtu* | -xenix* \
  	      | -hiux* | -386bsd* | -knetbsd* | -mirbsd* | -netbsd* \
diff -rcN src/binutils-org/gas/configure.tgt src/binutils/gas/configure.tgt
*** src/binutils-org/gas/configure.tgt	2012-09-04 14:53:45.000000000 +0200
--- src/binutils/gas/configure.tgt	2013-12-01 18:35:00.698391650 +0100
***************
*** 98,103 ****
--- 98,104 ----
  generic_target=${cpu_type}-$vendor-$os
  # Note: This table is alpha-sorted, please try to keep it that way.
  case ${generic_target} in
+   i386-*-escape*)           fmt=elf ;;
    aarch64*-*-elf)			fmt=elf;;
    aarch64*-*-linux*)			fmt=elf em=linux ;;
  
diff -rcN src/binutils-org/ld/configure.tgt src/binutils/ld/configure.tgt
*** src/binutils-org/ld/configure.tgt	2012-09-04 14:53:47.000000000 +0200
--- src/binutils/ld/configure.tgt	2013-12-01 18:35:22.414391477 +0100
***************
*** 27,32 ****
--- 27,33 ----
  # architecture variants should be kept together even if their names
  # break the alpha sorting.
  case "${targ}" in
+ i[3-7]86-*-escape*)    targ_emul=escape_i386 ;;
  aarch64_be-*-elf)	targ_emul=aarch64elfb
  			targ_extra_emuls="aarch64elf armelfb armelf" ;;
  aarch64-*-elf)		targ_emul=aarch64elf
diff -rcN src/binutils-org/ld/emulparams/escape_i386.sh src/binutils/ld/emulparams/escape_i386.sh
*** src/binutils-org/ld/emulparams/escape_i386.sh	1970-01-01 01:00:00.000000000 +0100
--- src/binutils/ld/emulparams/escape_i386.sh	2013-12-01 18:35:44.294391302 +0100
***************
*** 0 ****
--- 1,15 ----
+ SCRIPT_NAME=elf
+ OUTPUT_FORMAT=elf32-i386
+ TEXT_START_ADDR=0x00001000
+ MAXPAGESIZE="CONSTANT (MAXPAGESIZE)"
+ COMMONPAGESIZE="CONSTANT (COMMONPAGESIZE)"
+ ARCH=i386
+ MACHINE=
+ NOP=0x90909090
+ TEMPLATE_NAME=elf32
+ GENERATE_SHLIB_SCRIPT=yes
+ GENERATE_PIE_SCRIPT=yes
+ NO_SMALL_DATA=yes
+ SEPARATE_GOTPLT=12
+ DATA_ADDR="ALIGN(${MAXPAGESIZE})"
+ SHLIB_DATA_ADDR="ALIGN(${MAXPAGESIZE})"
diff -rcN src/binutils-org/ld/Makefile.in src/binutils/ld/Makefile.in
*** src/binutils-org/ld/Makefile.in	2013-03-25 09:06:23.000000000 +0100
--- src/binutils/ld/Makefile.in	2013-12-01 18:36:07.114391121 +0100
***************
*** 2134,2139 ****
--- 2134,2141 ----
  
  @TDIRS@
  
+ eescape_i386.c: $(srcdir)/emulparams/escape_i386.sh $(ELF_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
+ 	${GENSCRIPTS} escape_i386 "$(tdir_escape_i386)"
  eaix5ppc.c: $(srcdir)/emulparams/aix5ppc.sh \
    $(srcdir)/emultempl/aix.em $(srcdir)/scripttempl/aix.sc ${GEN_DEPENDS}
  	${GENSCRIPTS} aix5ppc "$(tdir_aixppc)"
