diff -rcN src/binutils-org/bfd/config.bfd src/binutils/bfd/config.bfd
*** src/binutils-org/bfd/config.bfd	2017-07-19 16:25:10.895137875 +0200
--- src/binutils/bfd/config.bfd	2017-07-19 16:32:42.842349588 +0200
***************
*** 668,673 ****
--- 668,677 ----
      targ_defvec=i386_elf32_vec
      targ_selvecs=iamcu_elf32_vec
      ;;
+   i[3-7]86-*-escape*)
+     targ_defvec=i386_elf32_vec
+     targ_selvecs=i386_elf32_vec
+     ;;
    i[3-7]86-*-dicos*)
      targ_defvec=i386_elf32_vec
      targ_selvecs=iamcu_elf32_vec
diff -rcN src/binutils-org/bfd/elf32-i386.c src/binutils/bfd/elf32-i386.c
*** src/binutils-org/bfd/elf32-i386.c	2017-07-19 16:25:10.898471188 +0200
--- src/binutils/bfd/elf32-i386.c	2017-07-19 16:32:42.869016094 +0200
***************
*** 533,539 ****
  /* The name of the dynamic interpreter.  This is put in the .interp
     section.  */
  
! #define ELF_DYNAMIC_INTERPRETER "/usr/lib/libc.so.1"
  
  /* If ELIMINATE_COPY_RELOCS is non-zero, the linker will try to avoid
     copying dynamic variables from a shared lib into an app's dynbss
--- 533,539 ----
  /* The name of the dynamic interpreter.  This is put in the .interp
     section.  */
  
! #define ELF_DYNAMIC_INTERPRETER "/bin/dynlink"
  
  /* If ELIMINATE_COPY_RELOCS is non-zero, the linker will try to avoid
     copying dynamic variables from a shared lib into an app's dynbss
diff -rcN src/binutils-org/config.sub src/binutils/config.sub
*** src/binutils-org/config.sub	2017-07-19 16:25:10.885137938 +0200
--- src/binutils/config.sub	2017-07-19 16:32:42.882349346 +0200
***************
*** 1386,1392 ****
  	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -auroraux* | -solaris* \
  	      | -sym* | -kopensolaris* | -plan9* \
  	      | -amigaos* | -amigados* | -msdos* | -newsos* | -unicos* | -aof* \
! 	      | -aos* | -aros* | -cloudabi* | -sortix* \
  	      | -nindy* | -vxsim* | -vxworks* | -ebmon* | -hms* | -mvs* \
  	      | -clix* | -riscos* | -uniplus* | -iris* | -rtu* | -xenix* \
  	      | -hiux* | -386bsd* | -knetbsd* | -mirbsd* | -netbsd* \
--- 1386,1392 ----
  	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -auroraux* | -solaris* \
  	      | -sym* | -kopensolaris* | -plan9* \
  	      | -amigaos* | -amigados* | -msdos* | -newsos* | -unicos* | -aof* \
! 	      | -aos* | -aros* | -cloudabi* | -sortix* | -escape* \
  	      | -nindy* | -vxsim* | -vxworks* | -ebmon* | -hms* | -mvs* \
  	      | -clix* | -riscos* | -uniplus* | -iris* | -rtu* | -xenix* \
  	      | -hiux* | -386bsd* | -knetbsd* | -mirbsd* | -netbsd* \
diff -rcN src/binutils-org/gas/configure.tgt src/binutils/gas/configure.tgt
*** src/binutils-org/gas/configure.tgt	2017-07-19 16:25:10.935137624 +0200
--- src/binutils/gas/configure.tgt	2017-07-19 16:32:42.892349286 +0200
***************
*** 122,127 ****
--- 122,128 ----
  generic_target=${cpu_type}-$vendor-$os
  # Note: This table is alpha-sorted, please try to keep it that way.
  case ${generic_target} in
+   i386-*-escape*)			fmt=elf;;
    aarch64*-*-elf | aarch64*-*-rtems*)	fmt=elf;;
    aarch64*-*-fuchsia*)			fmt=elf;;
    aarch64*-*-linux*)			fmt=elf em=linux ;;
diff -rcN src/binutils-org/ld/configure.tgt src/binutils/ld/configure.tgt
*** src/binutils-org/ld/configure.tgt	2017-07-19 16:25:11.131803057 +0200
--- src/binutils/ld/configure.tgt	2017-07-19 16:32:42.892349286 +0200
***************
*** 45,50 ****
--- 45,51 ----
  # architecture variants should be kept together even if their names
  # break the alpha sorting.
  case "${targ}" in
+ i[3-7]86-*-escape*)	targ_emul=escape_i386 ;;
  aarch64_be-*-elf)	targ_emul=aarch64elfb
  			targ_extra_emuls="aarch64elf aarch64elf32 aarch64elf32b armelfb armelf" ;;
  aarch64-*-elf | aarch64-*-rtems*)
diff -rcN src/binutils-org/ld/emulparams/escape_i386.sh src/binutils/ld/emulparams/escape_i386.sh
*** src/binutils-org/ld/emulparams/escape_i386.sh	1970-01-01 01:00:00.000000000 +0100
--- src/binutils/ld/emulparams/escape_i386.sh	2017-07-19 16:32:42.912349165 +0200
***************
*** 0 ****
--- 1,15 ----
+ SCRIPT_NAME=elf
+ OUTPUT_FORMAT=elf32-i386
+ TEXT_START_ADDR=0x00001000
+ MAXPAGESIZE="CONSTANT (MAXPAGESIZE)"
+ COMMONPAGESIZE="CONSTANT (COMMONPAGESIZE)"
+ ARCH=i386
+ MACHINE=
+ NOP=0x90909090
+ TEMPLATE_NAME=elf32
+ GENERATE_SHLIB_SCRIPT=yes
+ GENERATE_PIE_SCRIPT=yes
+ NO_SMALL_DATA=yes
+ SEPARATE_GOTPLT=12
+ DATA_ADDR="ALIGN(${MAXPAGESIZE})"
+ SHLIB_DATA_ADDR="ALIGN(${MAXPAGESIZE})"
diff -rcN src/binutils-org/ld/Makefile.in src/binutils/ld/Makefile.in
*** src/binutils-org/ld/Makefile.in	2017-07-19 16:25:11.138469682 +0200
--- src/binutils/ld/Makefile.in	2017-07-19 16:34:55.081552640 +0200
***************
*** 535,540 ****
--- 535,541 ----
  	eaixrs6.c \
  	ealpha.c \
  	ealphavms.c \
+ 	eescape_i386.c \
  	earcv2elf.c \
  	earcv2elfx.c \
  	earcelf.c \
***************
*** 2313,2318 ****
--- 2314,2322 ----
    $(srcdir)/emultempl/needrelax.em \
    $(srcdir)/scripttempl/alphavms.sc ${GEN_DEPENDS}
  
+ eescape_i386.c: $(srcdir)/emulparams/escape_i386.sh \
+   $(ELF_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
+ 
  earcv2elf.c: $(srcdir)/emulparams/arcv2elf.sh \
    $(ELF_DEPS) $(srcdir)/scripttempl/elfarcv2.sc ${GEN_DEPENDS}
  
