diff -rcN src/binutils-org/bfd/config.bfd src/binutils/bfd/config.bfd
*** src/binutils-org/bfd/config.bfd	2017-03-02 09:23:53.000000000 +0100
--- src/binutils/bfd/config.bfd	2017-07-20 16:11:01.092748357 +0200
***************
*** 239,244 ****
--- 239,249 ----
  
  # START OF targmatch.h
  #ifdef BFD64
+   x86_64-*-escape)
+     targ_defvec=x86_64_elf64_vec
+     targ_selvecs="i386_elf32_vec"
+     want64=true
+     ;;
    aarch64-*-darwin*)
      targ_defvec=aarch64_mach_o_vec
      targ_selvecs="arm_mach_o_vec mach_o_le_vec mach_o_be_vec mach_o_fat_vec"
diff -rcN src/binutils-org/bfd/elf32-i386.c src/binutils/bfd/elf32-i386.c
*** src/binutils-org/bfd/elf32-i386.c	2017-03-02 09:23:53.000000000 +0100
--- src/binutils/bfd/elf32-i386.c	2017-07-20 16:11:46.349167855 +0200
***************
*** 533,539 ****
  /* The name of the dynamic interpreter.  This is put in the .interp
     section.  */
  
! #define ELF_DYNAMIC_INTERPRETER "/usr/lib/libc.so.1"
  
  /* If ELIMINATE_COPY_RELOCS is non-zero, the linker will try to avoid
     copying dynamic variables from a shared lib into an app's dynbss
--- 533,539 ----
  /* The name of the dynamic interpreter.  This is put in the .interp
     section.  */
  
! #define ELF_DYNAMIC_INTERPRETER "/bin/dynlink"
  
  /* If ELIMINATE_COPY_RELOCS is non-zero, the linker will try to avoid
     copying dynamic variables from a shared lib into an app's dynbss
diff -rcN src/binutils-org/bfd/elf64-x86-64.c src/binutils/bfd/elf64-x86-64.c
*** src/binutils-org/bfd/elf64-x86-64.c	2017-03-02 09:23:53.000000000 +0100
--- src/binutils/bfd/elf64-x86-64.c	2017-07-20 16:11:29.019262485 +0200
***************
*** 531,538 ****
  /* The name of the dynamic interpreter.	 This is put in the .interp
     section.  */
  
! #define ELF64_DYNAMIC_INTERPRETER "/lib/ld64.so.1"
! #define ELF32_DYNAMIC_INTERPRETER "/lib/ldx32.so.1"
  
  /* If ELIMINATE_COPY_RELOCS is non-zero, the linker will try to avoid
     copying dynamic variables from a shared lib into an app's dynbss
--- 531,538 ----
  /* The name of the dynamic interpreter.	 This is put in the .interp
     section.  */
  
! #define ELF64_DYNAMIC_INTERPRETER "/bin/dynlink"
! #define ELF32_DYNAMIC_INTERPRETER "/bin/dynlink"
  
  /* If ELIMINATE_COPY_RELOCS is non-zero, the linker will try to avoid
     copying dynamic variables from a shared lib into an app's dynbss
diff -rcN src/binutils-org/config.sub src/binutils/config.sub
*** src/binutils-org/config.sub	2016-12-23 09:40:17.000000000 +0100
--- src/binutils/config.sub	2017-07-20 16:12:19.638986136 +0200
***************
*** 1384,1390 ****
  	-gnu* | -bsd* | -mach* | -minix* | -genix* | -ultrix* | -irix* \
  	      | -*vms* | -sco* | -esix* | -isc* | -aix* | -cnk* | -sunos | -sunos[34]*\
  	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -auroraux* | -solaris* \
! 	      | -sym* | -kopensolaris* | -plan9* \
  	      | -amigaos* | -amigados* | -msdos* | -newsos* | -unicos* | -aof* \
  	      | -aos* | -aros* | -cloudabi* | -sortix* \
  	      | -nindy* | -vxsim* | -vxworks* | -ebmon* | -hms* | -mvs* \
--- 1384,1390 ----
  	-gnu* | -bsd* | -mach* | -minix* | -genix* | -ultrix* | -irix* \
  	      | -*vms* | -sco* | -esix* | -isc* | -aix* | -cnk* | -sunos | -sunos[34]*\
  	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -auroraux* | -solaris* \
! 	      | -sym* | -kopensolaris* | -plan9* | -escape* \
  	      | -amigaos* | -amigados* | -msdos* | -newsos* | -unicos* | -aof* \
  	      | -aos* | -aros* | -cloudabi* | -sortix* \
  	      | -nindy* | -vxsim* | -vxworks* | -ebmon* | -hms* | -mvs* \
diff -rcN src/binutils-org/gas/configure.tgt src/binutils/gas/configure.tgt
*** src/binutils-org/gas/configure.tgt	2017-03-02 09:23:53.000000000 +0100
--- src/binutils/gas/configure.tgt	2017-07-20 16:13:20.745319440 +0200
***************
*** 122,127 ****
--- 122,129 ----
  generic_target=${cpu_type}-$vendor-$os
  # Note: This table is alpha-sorted, please try to keep it that way.
  case ${generic_target} in
+   i*86-*-escape*)			fmt=elf ;;
+   x86_64-*-escape*)			fmt=elf ;;
    aarch64*-*-elf | aarch64*-*-rtems*)	fmt=elf;;
    aarch64*-*-fuchsia*)			fmt=elf;;
    aarch64*-*-linux*)			fmt=elf em=linux ;;
diff -rcN src/binutils-org/ld/configure.tgt src/binutils/ld/configure.tgt
*** src/binutils-org/ld/configure.tgt	2017-03-02 09:23:54.000000000 +0100
--- src/binutils/ld/configure.tgt	2017-07-20 16:14:02.945089372 +0200
***************
*** 263,268 ****
--- 263,269 ----
  			targ64_extra_libpath=elf_x86_64
  			targ_extra_libpath=elf32_x86_64
  			tdir_i386linux=${targ_alias}aout ;;
+ x86_64-*-escape) targ_emul=escape_x86_64 ;;
  x86_64-*-linux-gnux32)	targ_emul=elf32_x86_64
  			targ_extra_emuls="elf_x86_64 elf_i386 elf_iamcu i386linux elf_l1om elf_k1om"
  			targ_extra_libpath="elf_i386 elf_iamcu elf_x86_64 elf_l1om elf_k1om"
diff -rcN src/binutils-org/ld/emulparams/escape_x86_64.sh src/binutils/ld/emulparams/escape_x86_64.sh
*** src/binutils-org/ld/emulparams/escape_x86_64.sh	1970-01-01 01:00:00.000000000 +0100
--- src/binutils/ld/emulparams/escape_x86_64.sh	2017-07-20 16:14:23.381644664 +0200
***************
*** 0 ****
--- 1,17 ----
+ SCRIPT_NAME=elf
+ ELFSIZE=64
+ OUTPUT_FORMAT=elf64-x86-64
+ TEXT_START_ADDR=0x00001000
+ MAXPAGESIZE="CONSTANT (MAXPAGESIZE)"
+ COMMONPAGESIZE="CONSTANT (COMMONPAGESIZE)"
+ ARCH="i386:x86-64"
+ MACHINE=
+ NOP=0x90909090
+ TEMPLATE_NAME=elf32
+ GENERATE_SHLIB_SCRIPT=yes
+ GENERATE_PIE_SCRIPT=yes
+ NO_SMALL_DATA=yes
+ SEPARATE_GOTPLT=24
+ DATA_ADDR="ALIGN(${MAXPAGESIZE})"
+ SHLIB_DATA_ADDR="ALIGN(${MAXPAGESIZE})"
+ 
diff -rcN src/binutils-org/ld/Makefile.in src/binutils/ld/Makefile.in
*** src/binutils-org/ld/Makefile.in	2017-03-02 09:23:54.000000000 +0100
--- src/binutils/ld/Makefile.in	2017-07-20 16:16:00.447782730 +0200
***************
*** 535,540 ****
--- 535,541 ----
  	eaixrs6.c \
  	ealpha.c \
  	ealphavms.c \
+ 	eescape_x86_64.c \
  	earcv2elf.c \
  	earcv2elfx.c \
  	earcelf.c \
***************
*** 2313,2318 ****
--- 2314,2322 ----
    $(srcdir)/emultempl/needrelax.em \
    $(srcdir)/scripttempl/alphavms.sc ${GEN_DEPENDS}
  
+ eescape_x86_64.c: $(srcdir)/emulparams/escape_x86_64.sh \
+   $(ELF_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
+ 
  earcv2elf.c: $(srcdir)/emulparams/arcv2elf.sh \
    $(ELF_DEPS) $(srcdir)/scripttempl/elfarcv2.sc ${GEN_DEPENDS}
  
