diff -rcN src/gcc-org/gcc/graphite.h src/gcc/gcc/graphite.h
--- src/gcc-org/gcc/graphite.h 2018-01-03 11:03:58.000000000 +0000
+++ src/gcc/gcc/graphite.h 2018-07-29 17:42:55.836121000 +0000
@@ -37,6 +37,8 @@
 #include <isl/schedule.h>
 #include <isl/ast_build.h>
 #include <isl/schedule_node.h>
+#include <isl/id.h>
+#include <isl/space.h>

 typedef struct poly_dr *poly_dr_p;

diff -rcN src/gcc-org/config.sub src/gcc/config.sub
*** src/gcc-org/config.sub	2017-07-20 16:19:40.923251029 +0200
--- src/gcc/config.sub	2017-07-20 16:20:53.579523458 +0200
***************
*** 1388,1394 ****
  	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -auroraux* | -solaris* \
  	      | -sym* | -kopensolaris* | -plan9* \
  	      | -amigaos* | -amigados* | -msdos* | -newsos* | -unicos* | -aof* \
! 	      | -aos* | -aros* | -cloudabi* | -sortix* \
  	      | -nindy* | -vxsim* | -vxworks* | -ebmon* | -hms* | -mvs* \
  	      | -clix* | -riscos* | -uniplus* | -iris* | -rtu* | -xenix* \
  	      | -hiux* | -386bsd* | -knetbsd* | -mirbsd* | -netbsd* \
--- 1388,1394 ----
  	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -auroraux* | -solaris* \
  	      | -sym* | -kopensolaris* | -plan9* \
  	      | -amigaos* | -amigados* | -msdos* | -newsos* | -unicos* | -aof* \
! 	      | -aos* | -aros* | -cloudabi* | -sortix* | -escape* \
  	      | -nindy* | -vxsim* | -vxworks* | -ebmon* | -hms* | -mvs* \
  	      | -clix* | -riscos* | -uniplus* | -iris* | -rtu* | -xenix* \
  	      | -hiux* | -386bsd* | -knetbsd* | -mirbsd* | -netbsd* \
diff -rcN src/gcc-org/gcc/config/escape.h src/gcc/gcc/config/escape.h
*** src/gcc-org/gcc/config/escape.h	1970-01-01 01:00:00.000000000 +0100
--- src/gcc/gcc/config/escape.h	2017-07-20 16:21:17.092729277 +0200
***************
*** 0 ****
--- 1,31 ----
+ #undef   STARTFILE_SPEC
+ #define STARTFILE_SPEC "%{shared|pie:crt0S.o%s;:crt0.o%s} %{shared:crt1S.o%s;:crt1.o%s} %{shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+ 
+ #undef   ENDFILE_SPEC
+ #define ENDFILE_SPEC "%{shared|pie:crtendS.o%s;:crtend.o%s} %{shared:crtnS.o%s;:crtn.o%s}"
+ 
+ #undef TARGET_OS_CPP_BUILTINS
+ #define TARGET_OS_CPP_BUILTINS()      \
+   do {                                \
+     /* to force a strcmp for typeinfo-compares */ \
+     builtin_define ("__GXX_MERGED_TYPEINFO_NAMES=0"); \
+     builtin_define ("__GXX_TYPEINFO_EQUALITY_INLINE=0"); \
+     builtin_define_std ("escape");      \
+     builtin_define_std ("unix");      \
+     builtin_assert ("system=escape");   \
+     builtin_assert ("system=unix");   \
+   } while(0);
+ 
+ #undef LINK_GCC_C_SEQUENCE_SPEC
+ #define LINK_GCC_C_SEQUENCE_SPEC \
+   "%{static:--start-group} %G %L %{static:--end-group}%{!static:%G}"
+ 
+ /* Use --as-needed -lgcc_s for eh support.  */
+ #ifdef HAVE_LD_AS_NEEDED
+ #define USE_LD_AS_NEEDED 1
+ #endif
+ 
+ #define NO_IMPLICIT_EXTERN_C
+ 
+ #define USE_TM_CLONE_REGISTRY 0
+ 
diff -rcN src/gcc-org/gcc/config.gcc src/gcc/gcc/config.gcc
*** src/gcc-org/gcc/config.gcc	2017-07-20 16:19:42.399909680 +0200
--- src/gcc/gcc/config.gcc	2017-07-20 16:22:30.335665623 +0200
***************
*** 615,620 ****
--- 615,625 ----
  
  # Common parts for widely ported systems.
  case ${target} in
+ *-*-escape*)
+   gas=yes
+   gnu_ld=yes
+   default_use_cxa_atexit=yes
+   ;;
  *-*-darwin*)
    tmake_file="t-darwin ${cpu_type}/t-darwin"
    tm_file="${tm_file} darwin.h"
***************
*** 918,923 ****
--- 923,933 ----
  esac
  
  case ${target} in
+ x86_64-*-escape*)
+ 	tm_file="${tm_file} i386/unix.h i386/att.h dbxelf.h elfos.h i386/i386elf.h i386/x86-64.h escape.h"
+ 	tmake_file="${tmake_file} t-svr4"
+ 	use_fixproto=yes
+ 	;;
  aarch64*-*-elf | aarch64*-*-fuchsia* | aarch64*-*-rtems*)
  	tm_file="${tm_file} dbxelf.h elfos.h newlib-stdint.h"
  	tm_file="${tm_file} aarch64/aarch64-elf.h aarch64/aarch64-elf-raw.h"
diff -rcN src/gcc-org/libgcc/config/t-escape src/gcc/libgcc/config/t-escape
*** src/gcc-org/libgcc/config/t-escape	1970-01-01 01:00:00.000000000 +0100
--- src/gcc/libgcc/config/t-escape	2017-07-20 16:23:43.521924579 +0200
***************
*** 0 ****
--- 1,5 ----
+ LIBGCC2_CFLAGS += -Wl,-z,max-page-size=0x1000
+ 
+ # for some reason we need -Wl,-shared as well, i.e. -shared is not enough
+ SHLIB_LDFLAGS = -Wl,-shared -Wl,-soname,$(SHLIB_SONAME)
+ 
diff -rcN src/gcc-org/libgcc/config.host src/gcc/libgcc/config.host
*** src/gcc-org/libgcc/config.host	2017-07-20 16:19:41.283249076 +0200
--- src/gcc/libgcc/config.host	2017-07-20 16:24:06.011797730 +0200
***************
*** 329,334 ****
--- 329,338 ----
  esac
  
  case ${host} in
+ x86_64-*-escape*)
+ 	extra_parts="crtbegin.o crtbeginS.o crtend.o crtendS.o"
+ 	tmake_file="${tmake_file} t-escape i386/t-crtstuff t-slibgcc t-libgcc-pic t-crtstuff-pic"
+ 	;;
  aarch64*-*-elf | aarch64*-*-rtems*)
  	extra_parts="$extra_parts crtbegin.o crtend.o crti.o crtn.o"
  	extra_parts="$extra_parts crtfastmath.o"
diff -rcN src/gcc-org/libgcc/gthr-posix.h src/gcc/libgcc/gthr-posix.h
*** src/gcc-org/libgcc/gthr-posix.h	2017-07-20 16:19:41.286582390 +0200
--- src/gcc/libgcc/gthr-posix.h	2017-07-20 16:24:31.654986475 +0200
***************
*** 33,38 ****
--- 33,39 ----
  #define __GTHREADS_CXX0X 1
  
  #include <pthread.h>
+ #include <sched.h>
  
  #if ((defined(_LIBOBJC) || defined(_LIBOBJC_WEAK)) \
       || !defined(_GTHREAD_USE_MUTEX_TIMEDLOCK))
diff -rcN src/gcc-org/libstdc++-v3/configure src/gcc/libstdc++-v3/configure
*** src/gcc-org/libstdc++-v3/configure	2017-07-20 16:19:41.726580002 +0200
--- src/gcc/libstdc++-v3/configure	2017-07-20 16:26:42.007585569 +0200
***************
*** 11145,11150 ****
--- 11145,11158 ----
    shlibpath_var=LD_LIBRARY_PATH
    ;;
  
+ escape*)
+   library_names_spec='${libname}${release}${shared_ext}$versuffix ${libname}${shared_ext}$versuffix'
+   need_lib_prefix=no
+   need_version=no
+   dynamic_linker='/bin/dynlink'
+   hardcode_into_libs=yes
+   ;;
+ 
  *)
    dynamic_linker=no
    ;;
***************
*** 13257,13262 ****
--- 13265,13274 ----
  	esac
        ;;
  
+       escape*)
+       	ld_shlibs_CXX=yes
+       	;;
+ 
        tandem*)
          case $cc_basename in
            NCC*)
***************
*** 14645,14650 ****
--- 14657,14670 ----
    dynamic_linker='GNU/Linux ld.so'
    ;;
  
+ escape*)
+   library_names_spec='${libname}${release}${shared_ext}$versuffix ${libname}${shared_ext}$versuffix'
+   need_lib_prefix=no
+   need_version=no
+   dynamic_linker='/bin/dynlink'
+   hardcode_into_libs=yes
+   ;;
+ 
  netbsd*)
    version_type=sunos
    need_lib_prefix=no
diff -rcN src/gcc-org/libstdc++-v3/libsupc++/eh_term_handler.cc src/gcc/libstdc++-v3/libsupc++/eh_term_handler.cc
*** src/gcc-org/libstdc++-v3/libsupc++/eh_term_handler.cc	2017-07-20 16:19:41.489914620 +0200
--- src/gcc/libstdc++-v3/libsupc++/eh_term_handler.cc	2017-07-20 16:26:58.937490297 +0200
***************
*** 32,38 ****
     --disable-libstdcxx-verbose and rebuilding the library.
     In a freestanding environment, we default to this latter approach.  */
  
! #if _GLIBCXX_HOSTED && _GLIBCXX_VERBOSE && __cpp_exceptions
  /* The current installed user handler.  */
  std::terminate_handler __cxxabiv1::__terminate_handler =
  	__gnu_cxx::__verbose_terminate_handler;
--- 32,38 ----
     --disable-libstdcxx-verbose and rebuilding the library.
     In a freestanding environment, we default to this latter approach.  */
  
! #if 1 // _GLIBCXX_HOSTED && _GLIBCXX_VERBOSE && __cpp_exceptions
  /* The current installed user handler.  */
  std::terminate_handler __cxxabiv1::__terminate_handler =
  	__gnu_cxx::__verbose_terminate_handler;
diff -rcN src/gcc-org/libstdc++-v3/libsupc++/vterminate.cc src/gcc/libstdc++-v3/libsupc++/vterminate.cc
*** src/gcc-org/libstdc++-v3/libsupc++/vterminate.cc	2017-07-20 16:19:41.489914620 +0200
--- src/gcc/libstdc++-v3/libsupc++/vterminate.cc	2017-07-20 16:27:32.307302572 +0200
***************
*** 24,35 ****
  
  #include <bits/c++config.h>
  
! #if _GLIBCXX_HOSTED
  #include <cstdlib>
  #include <exception>
  #include <bits/exception_defines.h>
  #include <cxxabi.h>
! # include <cstdio>
  
  using namespace std;
  using namespace abi;
--- 24,40 ----
  
  #include <bits/c++config.h>
  
! #if 1 // _GLIBCXX_HOSTED
  #include <cstdlib>
  #include <exception>
  #include <bits/exception_defines.h>
  #include <cxxabi.h>
! // don't include cstdio here because thats too much trouble
! //# include <cstdio>
! extern "C" void abort(void);
! extern "C" int fputs(const char*,void*);
! extern "C" void *free(void*);
! extern void *stderr;
  
  using namespace std;
  using namespace abi;
***************
*** 62,68 ****
  	  int status = -1;
  	  char *dem = 0;
  	  
! 	  dem = __cxa_demangle(name, 0, 0, &status);
  
  	  fputs("terminate called after throwing an instance of '", stderr);
  	  if (status == 0)
--- 67,73 ----
  	  int status = -1;
  	  char *dem = 0;
  	  
! 	  // dem = __cxa_demangle(name, 0, 0, &status);
  
  	  fputs("terminate called after throwing an instance of '", stderr);
  	  if (status == 0)
